src/data_utils.py

```diff
@@ -0,0 +1,167 @@
+from typing import Dict, Optional, Sequence
+import json
+from transformers import (
+    PreTrainedTokenizer,
+    PreTrainedModel,
+)
+import tqdm 
+import copy
+import logging 
+import os
+import torch
+from torch.utils.data import Dataset
+
+
+IGNORE_INDEX = -100
+PAD_TOKEN = "[PAD]"
+EOS_TOKEN = "</s>"
+BOS_TOKEN = "<s>"
+UNK_TOKEN = "<unk>"
+PROMPT_DICT = {
+    "prompt_input": (
+        "Below is an instruction that describes a task, paired with an input that provides further context. "
+        "Write a response that appropriately completes the request.\n\n"
+        "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:"
+    ),
+    "prompt_no_input": (
+        "Below is an instruction that describes a task. "
+        "Write a response that appropriately completes the request.\n\n"
+        "### Instruction:\n{instruction}\n\n### Response:"
+    ),
+}
+
+
+def resize_tokenizer_embedding(
+    new_tokens: Dict,
+    tokenizer: PreTrainedTokenizer,
+    model: PreTrainedModel,
+):
+    """
+    Resize emebdding and tokenizer.
+    """
+    num_new_tokens = tokenizer.add_special_tokens(new_tokens)
+    if num_new_tokens > 0:
+        model.resize_token_embeddings(len(tokenizer))
+        
+        # Assign new embeddings the average value
+        in_embeddings = model.get_input_embeddings().weight.data
+        out_embeddings = model.get_output_embeddings().weight.data
+        in_embeddings_avg = in_embeddings[ :-num_new_tokens].mean(dim=0, keepdim=True)
+        out_embeddings_avg = out_embeddings[ :-num_new_tokens].mean(dim=0, keepdim=True)
+        
+        in_embeddings[-num_new_tokens: ] = in_embeddings_avg
+        out_embeddings[-num_new_tokens: ] = out_embeddings_avg
+        
+
+def tokenize(string: Sequence[str], tokenizer: PreTrainedTokenizer) -> Dict:
+    """ Tokenizer a list of strings"""
+    
+    tokenized_list = [
+        tokenizer(
+            text,
+            return_tensors="pt",
+            padding="longest",
+            max_length=tokenizer.model_max_length,
+            truncation=True,
+        )
+        for text in string
+    ]
+    
+    input_ids = labels = [tokenized.input_ids[0] for tokenized in tokenized_list]
+    input_ids_lens = labels_lens = [
+        tokenized.input_ids.ne(tokenizer.pad_token_id).sum().item() for tokenized in tokenized_list
+    ]
+    return dict(
+        input_ids=input_ids,
+        labels=labels,
+        input_ids_lens=input_ids_lens,
+        labels_lens=labels_lens,
+    )
+    
+    
+def preprocess(
+    sources: Sequence[str],
+    targets: Sequence[str],
+    tokenizer: PreTrainedTokenizer,
+) -> Dict:
+    """Preprocess the data by tokenizing."""
+    
+    examples = [s + t for s, t in zip(sources, targets)]
+    if not os.path.exists("alpaca_tokenized.pt"):
```

Comment:
Dataset path is hard coded. This is not a good idea.

Reply:
Generally true...but should be fine within a toy program